### ISSUE TITLE
radicle-surf: error handling refactoring

### ISSUE THREAD

**Issue by keepsimple1** [radicle-dev/radicle-git#91]
Currently, `radicle-surf` defines separate error types for each module (sometimes even [multiple error types in inside a module](https://github.com/radicle-dev/radicle-git/blob/9be2c3d1400ffd059a2e2c24fcc87938affbcdd7/radicle-surf/src/fs.rs#L36-L62)). 

The upside is that each error tells where (which module or type) it is from. 

The downside is that the error chaining could lead to circular chains, for example: 

[`object::Error`](https://github.com/radicle-dev/radicle-git/blob/9be2c3d1400ffd059a2e2c24fcc87938affbcdd7/radicle-surf/src/object.rs#L35-L49) can chain / wrap transparently `fs::error::Directory` and `git::Error`. Now `fs::error::Directory` or `git::Error` cannot (not using too much tricks) chain `object::Error`, otherwise it will cause a circle. The compile would fail.

Another downside is that , IMO, exposing many (more than one) error types from a single crate would add unnecessary complexity to its API and usage.

I'd like to propose to define a single public error type for `radicle-surf` crate that uses detailed error names to tell where (which module or type) the error is from.  The current [`git::Error`](https://github.com/radicle-dev/radicle-git/blob/9be2c3d1400ffd059a2e2c24fcc87938affbcdd7/radicle-surf/src/git/repo.rs#L58) can form the basis of such a type.

As an example, the current error `fs::error::Directory::InvalidPath(String)` can be moved into `git::Error::DirectoryInvalidPath(String)`.  i.e. the new error name has `Directory` as its prefix to indicate the module.

Further, if we get rid of `git/` and promote its contents, this public error type would just be `crate::Error`.

**Comment by FintanH** [comment]
> [`object::Error`](https://github.com/radicle-dev/radicle-git/blob/9be2c3d1400ffd059a2e2c24fcc87938affbcdd7/radicle-surf/src/object.rs#L35-L49) can chain / wrap transparently `fs::error::Directory` and `git::Error`. Now `fs::error::Directory` or `git::Error` cannot (not using too much tricks) chain `object::Error`, otherwise it will cause a circle. The compile would fail.

Good news :) https://github.com/radicle-dev/radicle-git/pull/92

I'm not sure I favour this proposal, but I realise this has surfaced as this problem https://github.com/radicle-dev/radicle-git/blob/9be2c3d1400ffd059a2e2c24fcc87938affbcdd7/radicle-surf/src/fs.rs#L273-L279. The way I view this is that this friction is telling us that there is a code smell -- the friction is a good thing! So let's ask "why is this happening?" It looks like we're only using the `repo` for getting the `last_commit`. Why are we getting the `last_commit`? Is it necessary to do that for the `Directory` type? I wouldn't say so because it's not necessary information for a `Directory` to be a `Directory`, but if someone wanted to get the `last_commit` they could easily use the `Oid` of the `Directory` to compute it elsewhere. 

Long story short, I'd prefer we embrace this friction as a way to tell us that we might be doing something that is crossing logical boundaries. Otherwise, the single `Error` type won't tell us what kind of clean boundaries we want to create, it'll be our kitchen sink of errors that can happen.

Let me know what you think, and if you feel strongly about this then we can consider the God Mode Error.

**Comment by keepsimple1** [comment]
Thanks for pointing out the example. In the case above, I think there are two separate parts: 

1. Should we store `last_commit` in `Directory`, or should we compute it when needed? 

For this, I will opt for computing it when needed. Please see PR https://github.com/radicle-dev/radicle-git/pull/93 .

2. Should we have `error::Directory` and `error::File` types, or merge them into a common error type for the crate? 

Even when we extract `last_commit` out and only compute it when needed, there is still a challenge for the error type returned when `directory.last_commit()` fails.  I.e. the underlying issue still exists.  (We could remove `Directory::last_commit()` and ask the user call `Repo::last_commit()` instead, but I think that would miss the point).

**Comment by FintanH** [comment]
> Thanks for pointing out the example. In the case above, I think there are two separate parts:
> 
>     1. Should we store `last_commit` in `Directory`, or should we compute it when needed?
> 
> 
> For this, I will opt for computing it when needed. Please see PR #93 .
> 
>     2. Should we have `error::Directory` and `error::File` types, or merge them into a common error type for the crate?
> 
> 
> Even when we extract `last_commit` out and only compute it when needed, there is still a challenge for the error type returned when `directory.last_commit()` fails. I.e. the underlying issue still exists. (We could remove `Directory::last_commit()` and ask the user call `Repo::last_commit()` instead, but I think that would miss the point).

I think you might have missed my point about computing when needed. I'm saying we don't have a `last_commit`/`snapshot_id` field in `File` or `Directory` _at all_. If a user wants that information then they use `repo.last_commit` with the necessary parameters from `File`/`Directory`.

**Comment by keepsimple1** [comment]
Now the `last_commit` issue has been settled, we can go back to the error handling ;-).  

Here is an example how the current user deals with the multiple error types exposed by `radicle-surf`:  

https://github.com/radicle-dev/heartwood/blob/897d439f95b0081995145e478cbe5bc71cc4fcc4/radicle-httpd/src/api/error.rs#L41-L51

In this case, the user has to define multiple variants in its own error type to map to different error types from the same crate (`radicle-surf`). If we expose only 1 error type from `radicle-surf` without losing error info, then it will be a win for the users.

**Comment by FintanH** [comment]
> In this case, the user has to define multiple variants in their own error type to map to different error types from the same crate (`radicle-surf`). If we expose only 1 public error type from `radicle-surf` without losing error info, then I think it will be a win for the users.
> 
> (`object::Error` will no longer exist, but the general pattern will be the same).

I've said this before, I think that's completely fine to have multiple variants, in fact, it's great! One knows that this is exactly an error to do with handling `object`s and the other is to do with interacting with `git` directly. It's useful, cognitively, to organise modules, types, and errors into buckets telling you how things are organised, what's used where, and how things can go wrong.

Collapsing the error down into one error type saves a few lines of code -- I don't think that's a win, personally.

**Comment by keepsimple1** [comment]
Just for fun, I asked `ChatGPT` this question ;-). Here is the screenshot: 

![Rust-Lib-Errors-ChatGPT](https://user-images.githubusercontent.com/7699244/214934455-0ad74d8a-4be4-479f-b017-b2fd3608cdc8.jpg)

**Comment by keepsimple1** [comment]
I think we can make relative small changes to achieve:  a) a crate level `Error` which is based off the existing `repo::Error`.  b) with a small repo specific error just like other module level errors. 

I've opened PR https://github.com/radicle-dev/radicle-git/pull/102 for this.

**Comment by keepsimple1** [comment]
Thanks for merging the PR.
